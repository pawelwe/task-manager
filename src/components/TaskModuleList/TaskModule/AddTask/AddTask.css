.AddTask {
  padding-top: 1.2rem;
  border-top: 1px solid #dddddd; }
  .AddTask_btn {
    font-size: 1.4rem;
    font-weight: bold;
    color: crimson;
    background-color: transparent; }
  .AddTask_input {
    display: block;
    width: 100%;
    background-color: #ffffff;
    margin-right: 1rem;
    font-size: 1.2rem;
    font-weight: 100;
    padding: 0.6rem 1rem;
    border: 1px solid #dddddd;
    border-bottom: 0; }
    .AddTask_input::placeholder {
      color: #999999; }
    .AddTask_input:first-of-type {
      margin-top: 1rem; }
    .AddTask_input:last-of-type {
      border-bottom: 1px solid #dddddd;
      margin-bottom: 1rem; }
