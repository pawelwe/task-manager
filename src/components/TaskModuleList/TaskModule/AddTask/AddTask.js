import React, { Component } from 'react';
import classes from './AddTask.scss';
import Input from '../../../Inputs/Input';
import isEqual from 'lodash.isequal';

class AddTask extends Component {
  state = {
    editMode: false,
    formIsValid: false,
    editedTaskId: null,
    newTaskForm: {
      taskName: {
        elementType: 'text',
        config: {
          type: 'text',
          placeholder: 'Task name',
        },
        value: '',
        validation: {
          minLength: 3,
          maxLength: 15,
          required: true,
          errorMessage: 'Not valid name',
        },
        touched: false,
        valid: false,
      },
      priority: {
        elementType: 'number',
        config: {
          type: 'number',
          placeholder: 'Priority (1-10)',
        },
        value: '',
        validation: {
          required: true,
          errorMessage: 'Should be a number 1-10',
        },
        touched: false,
        valid: false,
      },
      expiration: {
        elementType: 'number',
        config: {
          type: 'number',
          placeholder: 'Expiration',
        },
        value: '',
        validation: {
          required: true,
          errorMessage: 'Should be a number',
        },
        touched: false,
        valid: false,
        width: 50,
      },
      timeFrame: {
        elementType: 'select',
        config: {
          type: 'select',
        },
        validation: {},
        options: [
          { value: 'seconds', display: 'Seconds' },
          { value: 'minutes', display: 'Minutes' },
          { value: 'hours', display: 'Hours' },
          { value: 'days', display: 'Days' },
        ],
        value: 'minutes',
        touched: true,
        valid: true,
        width: 50,
      },
    },
  };

  checkIfEditMode() {
    let editMode = false;
    let updatedForm, editedTaskId;

    this.props.tasks.forEach(formElement => {
      if (formElement.editMode && formElement.editMode === true) {
        editMode = true;
      }
    });

    if (editMode) {
      this.props.tasks.forEach(task => {
        if (task.editMode === true) {
          // TODO update to more generic one
          updatedForm = {
            ...this.state.newTaskForm,
            taskName: {
              ...this.state.newTaskForm.taskName,
              value: task.name,
            },
            priority: {
              ...this.state.newTaskForm.priority,
              value: task.priority,
            },
            expiration: {
              ...this.state.newTaskForm.expiration,
              value: task.expirationPeriod,
            },
            timeFrame: {
              ...this.state.newTaskForm.timeFrame,
              value: task.timeFrame,
            },
          };
          editedTaskId = task.id;
        }
      });

      this.setState({
        editMode: true,
        newTaskForm: updatedForm,
        editedTaskId: editedTaskId,
      });
    }
  }

  componentDidUpdate(prevProps, prevState) {
    if (!isEqual(prevProps.tasks, this.props.tasks)) {
      this.checkIfEditMode();
    }
    console.info('Add task did update');
  }

  checkValidity(value, rules) {
    let isValid = true;

    if (rules.required) {
      isValid = value.trim() !== '' && isValid;
    }

    if (rules.minLength) {
      isValid = value.length >= rules.minLength && isValid;
    }

    if (rules.maxLength) {
      isValid = value.length <= rules.maxLength && isValid;
    }
    return isValid;
  }

  handleInputChange = (e, inputId) => {
    const updatedForm = { ...this.state.newTaskForm };

    const updatedFormElement = {
      ...updatedForm[inputId],
    };

    updatedFormElement.value = e.target.value;
    updatedFormElement.valid = this.checkValidity(
      updatedFormElement.value,
      updatedFormElement.validation,
    );
    updatedFormElement.touched = true;
    updatedForm[inputId] = updatedFormElement;

    // Check whole form
    let formIsValid = true;

    for (let inputId in updatedForm) {
      formIsValid = updatedForm[inputId].valid && formIsValid;
    }

    console.log('All valid?', formIsValid);

    this.setState({
      newTaskForm: updatedForm,
      formIsValid: formIsValid,
    });
  };

  handleEnter = e => {
    if (e.key === 'Enter') {
      this.confirm.click();
    }
  };

  handleClearForm = form => {
    return Object.keys(form)
      .map(input => {
        return {
          [input]: {
            ...form[input],
            value:
              form[input].elementType !== 'select' ? '' : form[input].value,
            valid: form[input].elementType !== 'select' ? false : true,
            touched: false,
          },
        };
      })
      .reduce((result, item) => {
        // Convert to Object
        let key = Object.keys(item)[0];
        result[key] = item[key];
        return result;
      }, {});
  };

  handleAddNewTask = e => {
    e.preventDefault();

    if (!this.state.formIsValid) {
      this.props.toggleAlert('addTaskInputAlert', true);
      return;
    }

    this.props.addNewTask(
      this.props.moduleId,
      this.state.newTaskForm.taskName.value,
      this.state.newTaskForm.priority.value,
      this.state.newTaskForm.expiration.value,
      this.state.newTaskForm.timeFrame.value,
    );

    const formCopy = { ...this.state.newTaskForm };

    this.setState({
      formIsValid: false,
      newTaskForm: this.handleClearForm(formCopy),
    });
  };

  handleSaveEditedTask = e => {
    e.preventDefault();
    console.info('Saving edited task...');

    const formCopy = { ...this.state.newTaskForm };

    this.props.toggleTaskEditMode(
      this.props.moduleId,
      this.state.editedTaskId,
      false,
    );

    this.props.editTask(
      this.props.moduleId,
      this.state.editedTaskId,
      this.state.newTaskForm.taskName.value,
      this.state.newTaskForm.priority.value,
      this.state.newTaskForm.expiration.value,
      this.state.newTaskForm.timeFrame.value,
    );
    this.setState({
      editMode: false,
      editedTaskId: null,
      newTaskForm: this.handleClearForm(formCopy),
    });
  };

  handleExitEditMode = e => {
    console.log('Exit edit mode...');
    e.preventDefault();
    const formCopy = { ...this.state.newTaskForm };

    this.props.toggleTaskEditMode(
      this.props.moduleId,
      this.state.editedTaskId,
      false,
    );

    this.setState({
      editMode: false,
      editedTaskId: null,
      newTaskForm: this.handleClearForm(formCopy),
    });
  };

  handleUpdateCreationDate = () => {
    console.log('Creation updateted...');
    this.props.editTask(
      this.props.moduleId,
      this.state.editedTaskId,
      this.state.newTaskForm.taskName.value,
      this.state.newTaskForm.priority.value,
      this.state.newTaskForm.expiration.value,
      this.state.newTaskForm.timeFrame.value,
      true,
    );
  };

  renderAddTaskForm() {
    let newTaskForm = Object.keys(this.state.newTaskForm).map(inputKey => {
      return {
        id: inputKey,
        ...this.state.newTaskForm[inputKey],
      };
    });

    const renderConfirmButton = () => {
      if (!this.state.editMode) {
        return (
          <button
            id="addNewTask"
            // ref={node => (this.confirm = node)}
            onClick={this.handleAddNewTask}
            className={classes.AddTask_btn}
          >
            + Add New Task
          </button>
        );
      } else {
        return (
          <div className={classes.AddTask_toolBar}>
            <button
              id="addNewTask"
              onClick={this.handleSaveEditedTask}
              className={classes.AddTask_btn}
            >
              <span>✓ </span>
              Save Task
            </button>
            <span className={classes.AddTask_toolBar_separator} />
            <button
              id="addNewTask"
              onClick={this.handleExitEditMode}
              className={classes.AddTask_btn}
            >
              <span>✗ </span>
              Cancel
            </button>
          </div>
        );
      }
    };

    return (
      <form>
        {this.state.editMode && (
          <div
            data-title="Update creation date to current time"
            className={`${classes.AddTask_updateTime} tooltip`}
            onClick={this.handleUpdateCreationDate}
          >
            <span className="cursorPointer">♼</span>
          </div>
        )}
        {newTaskForm.map(formElement => (
          <Input
            key={formElement.id}
            className={classes['AddTask_' + formElement.elementType]}
            elementType={formElement.elementType}
            value={formElement.value}
            config={formElement.config}
            valid={formElement.valid}
            shouldValidate={formElement.validation}
            options={formElement.options}
            touched={formElement.touched}
            width={formElement.width ? formElement.width : 100}
            onChange={e => {
              this.handleInputChange(e, formElement.id);
              this.handleEnter(e);
            }}
          />
        ))}
        <ul className={classes.errorMessages}>
          {newTaskForm.map(formElement => {
            return (
              <li className={classes.errorMessages_item} key={formElement.id}>
                {!formElement.valid &&
                  formElement.touched && (
                    <p>☹ {formElement.validation.errorMessage}</p>
                  )}
              </li>
            );
          })}
        </ul>
        {renderConfirmButton()}
      </form>
    );
  }

  render() {
    return <div> {this.renderAddTaskForm()} </div>;
  }
}

export default AddTask;
