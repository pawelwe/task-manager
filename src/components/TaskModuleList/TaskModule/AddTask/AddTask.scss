.AddTask {
  padding-top: 1.2rem;
  border-top: 1px solid $medium-gray-3;

  &_btn {
    font-size: 1.4rem;
    font-weight: bold;
    color: $accent;
    background-color: transparent;
  }

  &_input {
    display: block;
    width: 100%;
    background-color: $white;
    margin-right: 1rem;
    font-size: 1.2rem;
    font-weight: 100;
    padding: 0.6rem 1rem;
    border: 1px solid $medium-gray-3;
    border-top: 0;
    &::placeholder {
      color: $medium-gray;
    }
    &:first-of-type {
      border-top: 1px solid $medium-gray-3;
      margin-top: 1rem;
    }
    &:last-of-type {
      border-bottom: 1px solid $medium-gray-3;
      margin-bottom: 1rem;
    }
  }
}

.errorMessages {
  margin-bottom: 1rem;
  &_item {
    font-size: 1rem;
    color: $invalid;
  }
}