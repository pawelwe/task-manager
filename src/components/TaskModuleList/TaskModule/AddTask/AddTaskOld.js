import React from 'react';
import classes from './AddTask.scss';
import Input from '../../../Inputs/Input';

const AddTask = ({ addNewTask, moduleId }) => {
  let taskName, expirationPeriod, priority, confirm;

  return (
    <footer>
      <Input
        elementType="text"
        placeholder="Task name"
        inputref={node => (taskName = node)}
        className={classes.AddTask_input}
      />
      <Input
        elementType="number"
        placeholder="Priority (1-10)"
        min="1"
        max="10"
        className={classes.AddTask_input}
        inputref={node => (priority = node)}
      />
      <Input
        elementType="number"
        min="1"
        max="24"
        placeholder="Expiration time (minutes)"
        className={classes.AddTask_input}
        inputref={node => (expirationPeriod = node)}
        onKeyPress={e => {
          if (e.key === 'Enter') {
            confirm.click();
          }
        }}
      />
      <button
        id="addNewTask"
        ref={node => (confirm = node)}
        onClick={() => {
          addNewTask(
            moduleId,
            taskName.value,
            priority.value,
            expirationPeriod.value,
          );
          if (taskName.value && priority.value && expirationPeriod.value) {
            taskName.value = priority.value = expirationPeriod.value = '';
          }
        }}
        className={classes.AddTask_btn}
      >
        + Add New Task
      </button>
    </footer>
  );
};

export default AddTask;
