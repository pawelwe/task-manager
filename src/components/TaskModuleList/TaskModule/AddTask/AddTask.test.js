import React from 'react';
import { mount } from 'enzyme';
import AddTask from './AddTask';

describe('AddTask Component', () => {
  const addNewTask = jest.fn();
  const wrapper = mount(<AddTask addNewTask={addNewTask} moduleId={1} />);

  it('Snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('addNewTask called', () => {
    wrapper.find('#addNewTask').simulate('click');
    expect(addNewTask).toHaveBeenCalled();
  });

});
