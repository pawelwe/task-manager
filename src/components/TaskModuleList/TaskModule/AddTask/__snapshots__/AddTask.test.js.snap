// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTask Component Snapshot 1`] = `
ReactWrapper {
  "length": 1,
  Symbol(enzyme.__unrendered__): <AddTask
    addNewTask={[Function]}
    moduleId={1}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "addNewTask": [Function],
      "moduleId": 1,
    },
    "ref": null,
    "rendered": Object {
      "instance": <footer>
        <input
                placeholder="Task name"
        />
        <input
                type="number"
                min="1"
                max="10"
                placeholder="Priority (1-10)"
        />
        <input
                type="number"
                min="1"
                max="24"
                placeholder="Expiration date (minutes)"
        />
        <button
                id="addNewTask"
        >
                + Add New Task
        </button>
</footer>,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            className={undefined}
            placeholder="Task name"
/>,
          <input
            className={undefined}
            max="10"
            min="1"
            placeholder="Priority (1-10)"
            type="number"
/>,
          <input
            className={undefined}
            max="24"
            min="1"
            onKeyPress={[Function]}
            placeholder="Expiration date (minutes)"
            type="number"
/>,
          <button
            className={undefined}
            id="addNewTask"
            onClick={[Function]}
>
            + Add New Task
</button>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": <input
            placeholder="Task name"
/>,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": undefined,
            "placeholder": "Task name",
          },
          "ref": [Function],
          "rendered": Array [
            undefined,
          ],
          "type": "input",
        },
        Object {
          "instance": <input
            type="number"
            min="1"
            max="10"
            placeholder="Priority (1-10)"
/>,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": undefined,
            "max": "10",
            "min": "1",
            "placeholder": "Priority (1-10)",
            "type": "number",
          },
          "ref": [Function],
          "rendered": Array [
            undefined,
          ],
          "type": "input",
        },
        Object {
          "instance": <input
            type="number"
            min="1"
            max="24"
            placeholder="Expiration date (minutes)"
/>,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": undefined,
            "max": "24",
            "min": "1",
            "onKeyPress": [Function],
            "placeholder": "Expiration date (minutes)",
            "type": "number",
          },
          "ref": [Function],
          "rendered": Array [
            undefined,
          ],
          "type": "input",
        },
        Object {
          "instance": <button
            id="addNewTask"
>
            + Add New Task
</button>,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "+ Add New Task",
            "className": undefined,
            "id": "addNewTask",
            "onClick": [Function],
          },
          "ref": [Function],
          "rendered": Array [
            "+ Add New Task",
          ],
          "type": "button",
        },
      ],
      "type": "footer",
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "addNewTask": [Function],
        "moduleId": 1,
      },
      "ref": null,
      "rendered": Object {
        "instance": <footer>
          <input
                    placeholder="Task name"
          />
          <input
                    type="number"
                    min="1"
                    max="10"
                    placeholder="Priority (1-10)"
          />
          <input
                    type="number"
                    min="1"
                    max="24"
                    placeholder="Expiration date (minutes)"
          />
          <button
                    id="addNewTask"
          >
                    + Add New Task
          </button>
</footer>,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              className={undefined}
              placeholder="Task name"
/>,
            <input
              className={undefined}
              max="10"
              min="1"
              placeholder="Priority (1-10)"
              type="number"
/>,
            <input
              className={undefined}
              max="24"
              min="1"
              onKeyPress={[Function]}
              placeholder="Expiration date (minutes)"
              type="number"
/>,
            <button
              className={undefined}
              id="addNewTask"
              onClick={[Function]}
>
              + Add New Task
</button>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": <input
              placeholder="Task name"
/>,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": undefined,
              "placeholder": "Task name",
            },
            "ref": [Function],
            "rendered": Array [
              undefined,
            ],
            "type": "input",
          },
          Object {
            "instance": <input
              type="number"
              min="1"
              max="10"
              placeholder="Priority (1-10)"
/>,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": undefined,
              "max": "10",
              "min": "1",
              "placeholder": "Priority (1-10)",
              "type": "number",
            },
            "ref": [Function],
            "rendered": Array [
              undefined,
            ],
            "type": "input",
          },
          Object {
            "instance": <input
              type="number"
              min="1"
              max="24"
              placeholder="Expiration date (minutes)"
/>,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": undefined,
              "max": "24",
              "min": "1",
              "onKeyPress": [Function],
              "placeholder": "Expiration date (minutes)",
              "type": "number",
            },
            "ref": [Function],
            "rendered": Array [
              undefined,
            ],
            "type": "input",
          },
          Object {
            "instance": <button
              id="addNewTask"
>
              + Add New Task
</button>,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "+ Add New Task",
              "className": undefined,
              "id": "addNewTask",
              "onClick": [Function],
            },
            "ref": [Function],
            "rendered": Array [
              "+ Add New Task",
            ],
            "type": "button",
          },
        ],
        "type": "footer",
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
